### PR TITLE
Deprecated Property Warning on Case Management Page

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

If a form question is being saved to a case property that has been deprecated in the DD then the following banner message will be shown on the Case Management/User Properties page:
Screenshot from 2024-04-12 11-45-56

Additionally, each row that is being saved to a deprecated case property will have a warning:
Screenshot from 2024-04-12 11-46-09

This is to help better notify users as to which case properties should be considered for replacement.

If a row contains an error, then the deprecated warning is hidden to prioritise showing the error instead:
Screenshot from 2024-04-12 11-47-23

## Technical Summary

Link to ticket [here](https://dimagi.atlassian.net/browse/SC-3376).

## Feature Flag

None

### Safety story

- Local testing
- QA done
- Unit tests exist

### Automated test coverage

Unit tests exist for the DD util functions that retrieve the case property description and deprecated status mappings.

### QA Plan

QA has been completed. Ticket is available [here](https://dimagi.atlassian.net/browse/QA-6355).

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by esoergel** [comment]
Haven't reviewed the code, but the feature is very cool!

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js [hunk 1]
```diff
@@ -503,6 +503,16 @@ hqDefine('app_manager/js/forms/case_config_ui', function () {
                     return self.case_transaction.case_type();
                 });
                 self.updatedDescription = ko.observable();
+                self.isDeprecated = ko.computed(function () {
+                    const config = self.case_transaction.caseConfig;
+                    if (self.key() !== 'name' && _(config.deprecatedPropertiesDict).has(self.caseType())) {
```
Comment (Charl1996) [inline corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:517]: I presume `self.key()` is the property name?

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:517]: That is correct.

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:517]: is the `has` check required? Does not it return undefined if the key is not present?

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js [hunk 2]
```diff
@@ -412,6 +412,15 @@ hqDefine('app_manager/js/forms/case_config_ui', function () {
                 return count;
             });
 
+            self.hasDeprecatedProperties = ko.computed(function () {
+                for (const p of self.case_properties()) {
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:416]: where is this `case_properties` set?
Also, we can't use the deprecated dictionary we created for this?

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:416]: > where is this case_properties set?

This is getting set [here](https://github.com/dimagi/commcare-hq/blob/d57aa086c36b5631f0d1f23cbcbea10e9ce0da78/corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js#L284) in `baseMapping` which gets used by both the `caseTransaction` and `usercaseTransaction` models for the case and user property pages, respectively. These models get instantiated [here](https://github.com/dimagi/commcare-hq/blob/d57aa086c36b5631f0d1f23cbcbea10e9ce0da78/corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js#L631) and [here](https://github.com/dimagi/commcare-hq/blob/d57aa086c36b5631f0d1f23cbcbea10e9ce0da78/corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js#L709).

> we can't use the deprecated dictionary we created for this?

The deprecated dictionary contains a mapping of all the project's deprecated properties. We need to see whether any of these are being currently used to map form questions to, which is why we look through `case_properties`.

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:517]: That is a good point, this isn't strictly necessary. It would return `undefined` which would ultimately cause the next conditional to be falsey, causing the function to correctly return false.

I don't see any harm in removing this check, so will be taking it out.

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: [d57981d](https://github.com/dimagi/commcare-hq/pull/34425/commits/d57981d5daf49bb502441e80c625851f04f80abb)

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: Cool. I see we did it for the first check. We can't do it for the following one as well?

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js (hunk 2 above)
Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:416]: Okay :+1:

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: Removing the second one opens the scenario where we might try and access a property from `depProps` when it is `undefined`, causing a `TypeError` exception to be thrown. 

Still, I agree we can simplify this a little. I'll remove the `has` check and replace it with `if (depProps && self.key() !== 'name')`

**Comment by zandre-eng** [comment]
Noting that on this PR the Javascript Lint test is failing due to 79 pre-existing linting problems in the `case_config_ui.js` file. Given the extent of the linting changes for this file, I will not be addressing them in this PR.

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js [hunk 3]
```diff
@@ -502,6 +512,14 @@ hqDefine('app_manager/js/forms/case_config_ui', function () {
                     return self.case_transaction.case_type();
                 });
                 self.updatedDescription = ko.observable();
+                self.isDeprecated = ko.computed(function () {
+                    const config = self.case_transaction.caseConfig;
+                    const depProps = config.deprecatedPropertiesDict[self.caseType()];
+                    if (depProps && self.key() !== 'name') {
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:518]: Nice!

**Review by mkangia — APPROVED** [review]

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js [hunk 4]
```diff
@@ -727,6 +727,7 @@ hqDefine('app_manager/js/forms/case_config_ui', function () {
                 return usercaseTransaction({
                     case_properties: case_properties,
                     case_preload: case_preload,
+                    case_type: 'commcare-user', // will get overridden by the default. Set here to check deprecated status for saved properties on initial page load
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:730]: @zandre-eng 
Can you tell more about why this change was made?
Is usercaseTransaction case_type actualy 'commcare-user'?

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:730]: Unlike the properties being saved under "Case Management" which go to whichever case type was selected in the case list, the properties from "User Properties" only get saved to the `commcare-user` case type. Since it's a constant, I thought it would be safe to put here as well. 

The reason we need it here is that we need `case_type` to be pre-defined so that we can do the deprecation check on the initial page load (which happens before the `usercaseTransaction` model has a chance to define its own `case_type` property.

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:730]: Can we do an alternate approach here?
I am not sure about populating this field with a constant so as to satisfy a validation check.
Is the validation check not relevant for `usercaseTransaction`? Can we skip that check for it then since we know it would not have a value set for case_type.

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:730]: > Can we skip that check for it then since we know it would not have a value set for case_type.

Do you mean skip showing a deprecated warning for case properties in the "User Properties" tab? It would be nice if we can still do the check for user properties. Users are able to deprecate user properties in the DD, so there could potentially be deprecated case properties that are selected. It would keep the UI consistent with the "Case Management" tab.

We do use the constant in other places as well, such as [here](https://github.com/dimagi/commcare-hq/blob/b23df94313306caf3c81826993938fa96dedb35f/corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js#L487) and [here](https://github.com/dimagi/commcare-hq/blob/b23df94313306caf3c81826993938fa96dedb35f/corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js#L721). Since this constant string is already being used I wonder then if it might be sufficient to have a global constant `COMMCARE_USER_CASE_TYPE` and then simply refer this throughout the file? That way, if it ever changes we only have to do it in one place in the file.

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:730]: is commcare-user the expected case type for usercaseTransaction but just set [later](https://github.com/dimagi/commcare-hq/blob/b23df94313306caf3c81826993938fa96dedb35f/corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js#L487) instead?

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:730]: That is correct. For case properties, the case type gets set dynamically [here](https://github.com/dimagi/commcare-hq/blob/b23df94313306caf3c81826993938fa96dedb35f/corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js#L462), whereas user properties have a hard-coded case type set where you've linked in your comment.

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:730]: okay, thank you :+1:

**Review by ajeety4 — APPROVED** [review]

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js [hunk 5]
```diff
@@ -1,5 +1,7 @@
 "use strict";
-hqDefine('app_manager/js/forms/case_config_ui', function () {
+hqDefine('app_manager/js/forms/case_config_ui', [
+    'hqwebapp/js/privileges',
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: Does this load all privileges of the domain to have them available in the UI?
If yes, would it better to just add to the context if the domain has access to DD and add that as initial page data?

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: The `hasPrivilege()` function which gets called [here](https://github.com/dimagi/commcare-hq/blame/4218209b82c61e68b3e329d470c04154eb250515/corehq/apps/hqwebapp/static/hqwebapp/js/privileges.js#L14) looks through `initialPageData`. A list of all the privileges are already available in JS and we're simply seeing if a specific one exists in the list.

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: Cool, thank you.

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js [hunk 6]
```diff
@@ -1,9 +1,10 @@
+"use strict";
 hqDefine('app_manager/js/forms/case_config_ui', function () {
-    "use strict";
     $(function () {
         var caseConfigUtils = hqImport('app_manager/js/case_config_utils'),
             initial_page_data = hqImport("hqwebapp/js/initial_page_data").get,
-            privileges = initial_page_data('add_ons_privileges'),
+            addOnsPrivileges = initial_page_data('add_ons_addOnsPrivileges'),
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:6]: where has the name changed for the initial page data variable?

File: corehq/apps/data_dictionary/util.py [hunk 7]
```diff
@@ -151,6 +151,25 @@ def get_case_property_label_dict(domain):
...
+    This returns a dictionary of the structure
+    {
+        case_type: {
+            case_property: is_deprecated,
+            ...
+        },
+        ...
+    }
+    for each case type and case property in the domain
+    """
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/data_dictionary/util.py:165]: > Thinking if we should consider just retaining a list of deprecated properties for each case type instead.

Do you mean instead of having the format of:
```
{
    case_type: {
        case_prop: is_deprecated,
        ...
    }
    ...
}
```

we instead do this:
```
{
    case_type: [
        case_prop
        ...
    ]
    ...
}
```

My only concern with the above approach is that while it is less data sent to the browser, it might be slower on the front-end to identify whether a case property is deprecated as we'll have to iterate through a list instead of directly looking up the case property through a dictionary.

> Additionally, in app manager, I believe we usually deal with one case type at a time? Though not sure if we should worry about that here.

This is true. We send what the active case type is [here](https://github.com/dimagi/commcare-hq/blob/cdd228d7871e83ec7c90bf8a5cd80eb1c97d5f28/corehq/apps/app_manager/views/forms.py#L789), so I'm wondering if, instead of changing the format from a dict to a list, we just send the data for the active case type.

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js (hunk 6 above)
Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:6]: This was renamed to `addOnsPrivileges` so as to not conflict with the privileges variable added in a followup commit (6afeec28e5cbbf70277ceec4f9edd3b4430ed476).

File: corehq/apps/data_dictionary/util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/data_dictionary/util.py:165]: > it might be slower on the front-end to identify whether a case property is deprecated as we'll have to iterate through a list instead of directly looking up the case property through a dictionary.

Absolutely. I was thinking about the same as well & was thinking that that change won't be noticeable. If any case type that has lots of case properties, if list lookup makes considerable difference, you could retain the dictionary.

> we just send the data for the active case type

Okay. I was not sure that we won't need other case types on that page. I was wondering if any related cases are referred?

Anyhow, this is not a blocking comment. I was just thinking about different ways here and thought a list would be more efficient. I leave the final call to you.

File: corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js (hunk 6 above)
Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:6]: You mean `add_ons_addOnsPrivileges`? I don't see a change to the naming of that variable.

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js:6]: Here's a history of how the variable was changed:
1. The `privileges` variable was renamed to `addOnsPrivileges` in commit [14afe79](https://github.com/dimagi/commcare-hq/pull/34425/commits/14afe79e763813c6d99201cf412daf4f7f044e7a) . 
2. Thereafter, `hqwebapp/js/privileges` was imported as `privileges` in commit [8c6d7a6](https://github.com/dimagi/commcare-hq/pull/34425/commits/8c6d7a6136c2eac974003157e1089df24886c927). 
3. This was then changed to a non-requireJS import in [6afeec2](https://github.com/dimagi/commcare-hq/pull/34425/commits/6afeec28e5cbbf70277ceec4f9edd3b4430ed476).

The import for `add_ons_privileges` was renamed to `addOnsPrivileges` and a new import for `hqwebapp/js/privileges` was added and called `privileges`.

Please let me know if the above still doesn't make sense and I'd be happy to take this offline if needed.

Comment (zandre-eng) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: After an offline with Manish, there was a misunderstanding from my side. The file path of the import was changed accidentally with a find and replace search. I've pushed a commit to fix the incorrect import path [92cd18d](https://github.com/dimagi/commcare-hq/pull/34425/commits/92cd18df13d51598cb8ad8f2012e3357d6878afc).

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/forms/case_config_ui.js]: :+1:

**Review by mkangia — COMMENTED** [review]
Looks good.
@zandre-eng Will just await your confirmation about a final testing after the last couple of changes.

File: corehq/apps/data_dictionary/util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/data_dictionary/util.py:166]: [5cc38da](https://github.com/dimagi/commcare-hq/pull/34425/commits/5cc38daba458a30b51eb79571eda701ed9ab0e01)

**Comment by zandre-eng** [comment]
> Will just await your confirmation about a final testing after the last couple of changes.

@mkangia Final changes have been done and I have done a final round of testing (including subcases) and can confirm that everything works as expected.

File: corehq/apps/data_dictionary/util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/data_dictionary/util.py:166]: > Since other data, such as the case property descriptions, are sending data for each case type I'll rather be safe and keep it the same for deprecated case properties

Ah, good detail :+1: 

Thanks!

**Review by mkangia — APPROVED** [review]
